lib/WebService/Async/Onfido/RateLimiter.pm

```diff
@@ -0,0 +1,152 @@
+package WebService::Async::Onfido::RateLimiter;
+
+use strict;
+use warnings;
+
+our $VERSION = '0.001';
+
+=head1 NAME
+
+WebService::Async::Onfido::RateLimiter - Module abstract
+
+=head1 SYNOPSIS
+
+    use WebService::Async::Onfido::RateLimiter;
+    my $instance = WebService::Async::Onfido::RateLimiter->new;
+
+=head1 DESCRIPTION
+
+=cut
+
+=head1 METHODS
+
+=cut
+
+use Carp qw();
+use Future;
+use mro;
+
+use parent qw(IO::Async::Notifier);
+
+=head1 NAME
+
+WebService::Async::CustomerIO::RateLimitter - This class provide possibility to limit amount
+of request in time interval
+
+=head1 SYNOPSIS
+
+=head1 DESCRIPTION
+
+=cut
+
+sub _init {
+    my ($self, $args) = @_;
+    for my $k (qw(limit interval)) {
+        die "Missing required argument: $k" unless exists $args->{$k};
+        die "Invalid value for $k: $args->{$k}" unless int($args->{$k}) > 0;
+        $self->{$k} = delete $args->{$k} if exists $args->{$k};
+    }
+
+    $self->{queue}   = [];
+    $self->{counter} = 0;
+
+    return $self->next::method($args);
+}
+
+=head2 interval
+
+=cut
+
+sub interval { return shift->{interval} }
+
+=head2 limit
+
+=cut
+
+sub limit { return shift->{limit} }
+
+=head2 acquire
+
+Method checks availability for free slot.
+It returns future, when slot will be available, then future will be resolved.
+
+=cut
+
+sub acquire {
+    my ($self) = @_;
+
+    $self->_start_timer;
+    return Future->done if ++$self->{counter} <= $self->limit;
+
+    my $current = $self->_current_queue;
+    $current->{counter}++;
+    return $current->{future};
+}
+
+=head2 is_limited
+
+=cut
+
+sub is_limited {
+    my ($self) = @_;
+    return $self->{counter} >= $self->limit;
+}
+
+sub _current_queue {
+    my ($self) = @_;
+
+    # +1 for getting correct position for edge cases like: limit 2, counter 4, should be 0
+    my $pos = int(($self->{counter} - ($self->limit + 1)) / $self->limit);
+
+    $self->{queue}[$pos] //= {
+        future  => $self->loop->new_future,
+        counter => 0
+    };
+
+    return $self->{queue}[$pos];
+}
+
+sub _start_timer {
```

Comment:
Should just be `timer`, no `_start_` prefix - it doesn't *start* the timer, it just creates one if it didn't already exist.